src/java.base/share/classes/java/lang/StackFrameInfo.java

```diff
@@ -66,26 +60,48 @@ public String getClassName() {
         return declaringClass().getName();
     }
 
-    @Override
-    public Class<?> getDeclaringClass() {
-        ensureRetainClassRefEnabled();
-        return declaringClass();
-    }
-
     @Override
     public String getMethodName() {
-        return JLIA.getName(memberName);
+        if (name == null) {
+            expandStackFrameInfo();
+            assert name != null;
+        }
+        return name;
     }
 
     @Override
     public MethodType getMethodType() {
         ensureRetainClassRefEnabled();
-        return JLIA.getMethodType(memberName);
+
+        if (type == null) {
+            expandStackFrameInfo();
+            assert type != null;
+        }
+
+        {
+            // Get a snapshot of type which doesn't get changed by racing threads.
+            final Object type = this.type;
+            if (type instanceof MethodType mt) {
+                return mt;
+            }
+        }
+
+        // type is not a MethodType yet.  Convert it thread-safely.
+        synchronized (this) {
+            if (type instanceof String sig) {
+                type = JLIA.getMethodType(sig, declaringClass().getClassLoader());
+            }
+            assert type instanceof MethodType : "bad method type " + type;
```

Comment:
Not sure this assert is useful given the next instruction is a cast to MethodType